### DISCUSSION TITLE
Support to facilitate remote sync

### DISCUSSION THREAD

**Discussion by JCKodel** [simolus3/drift#2457] [Q&A]
I started this thread on SQLite Forum: https://sqlite.org/forum/forumpost/cfe7bae7b0 asking if it was possible for SQLite to return data as JSON the same way MSSQL does.

My point was to facilitate the copy of inserted/updated/deleted data into a log table for later remote sync, using generic triggers instead of having to declare verbose known-columns queries.

Although they complete ignored the JSON output idea =P, they pointed out this feature: https://sqlite.org/sessionintro.html#obtaining_the_session_extension

Someone knows the best and easiest way to turn on that feature, on both Android and iOS?

The idea here is to create a transparent-as-possible system to eventually sync databases to a remote server.

**Comment by davidmartos96** [comment]
@JCKodel You might find https://electric-sql.com/ useful, it's currently in preview stage.

I'm maintaining an unofficial Dart client to be able to use it in Flutter applications.

https://github.com/SkillDevs/electric_dart

It transparently syncs SQLite and Postgres so you can write regular SQL operations.

**Reply by JCKodel** [reply]
> @JCKodel You might find https://electric-sql.com/ useful, it's currently in preview stage.
> 
Thank you, but Postgres support for non-English languages and JSON is dreadful. I use MSSQL only.

Also, that doesn't even support Flutter (and certainly not Drift).

**Reply by davidmartos96** [reply]
> Also, that doesn't even support Flutter (and certainly not Drift).

That's the second link I posted. There is a drift driver which works with stream queries

**Comment by simolus3** [comment]
> Someone knows the best and easiest way to turn on that feature, on both Android and iOS?

Stop depending on `sqlite3_flutter_libs` and compile sqlite3 yourself :)

On iOS, you pretty much just need to add `s.dependency 'sqlite3/session'` to your podspec.
On Android, you'd have to compile sqlite3 with the compile-time option mentioned in the link you've posted. You can take the [build scripts here](https://github.com/simolus3/sqlite-native-libraries/tree/master/sqlite3-native-library) as an inspiration and put them into your Android project. The option would have to be added [in this file](https://github.com/simolus3/sqlite-native-libraries/blob/bfa14a1d96e5bb05d04703dfd0448379ba6f92b1/sqlite3-native-library/cpp/CMakeLists.txt#L22-L43).

You could possibly also use [drift's runtime schema inspection](https://drift.simonbinder.eu/docs/advanced-features/schema_inspection/) APIs to iterate over all tables in a migration to generate a "log" table with triggers dynamically.

**Reply by JCKodel** [reply]
> You could possibly also use [drift's runtime schema inspection](https://drift.simonbinder.eu/docs/advanced-features/schema_inspection/) APIs to iterate over all tables in a migration to generate a "log" table with triggers dynamically.
>

While trying to do that, I found a missing feature that I need: how to know when a column references another table (i.e.: a column is a foreign key). 

I'm creating indexes and triggers when the database is created, and I want to create indexes for foreign keys (since they often appear in WHERE clauses).
